src/main/resources/assets/js/app/project/list/ProjectList.ts

```diff
@@ -1,16 +1,90 @@
 import {ListBox} from 'lib-admin-ui/ui/selector/list/ListBox';
 import {Project} from '../../settings/data/project/Project';
 import {ProjectListItem} from './ProjectListItem';
+import {ProjectHelper} from '../../settings/data/project/ProjectHelper';
 
 export class ProjectList
     extends ListBox<Project> {
 
+    private projectLevel: Map<string, number> = new Map<string, number>();
+    private projectChildren: Map<string, Project[]> = new Map<string, Project[]>();
+
     constructor() {
         super('project-list-items');
     }
 
     protected createItemView(item: Project, readOnly: boolean): ProjectListItem {
-        return new ProjectListItem(item);
+        const itemView: ProjectListItem = new ProjectListItem(item);
+        itemView.addClass(`level-${this.projectLevel.get(item.getName())}`);
+
+        return itemView;
+    }
+
+    setItems(items: Project[], silent?: boolean) {
+        this.buildProjectLevelsMap(items);
+        this.buildChildrenMap(items);
+        super.setItems(this.sortProjects(items), silent);
+    }
+
+    private buildProjectLevelsMap(items: Project[]) {
+        items.forEach((item: Project) => this.calcProjectLevel(item, items));
+    }
+
+    private calcProjectLevel(item: Project, items: Project[]) {
+        let parentName: string = item.getParent();
+        let level: number = 0;
+
+        while (parentName) {
+            const parent: Project = items.find((p: Project) => p.getName() === parentName);
+
+            if (parent) {
+                level++;
+                parentName = parent.getParent();
+            } else {
+                parentName = null;
+            }
+        }
+
+        this.projectLevel.set(item.getName(), level);
+    }
+
+    private buildChildrenMap(items: Project[]) {
+        items.forEach((item: Project) => this.projectChildren.set(item.getName(), this.getDirectProjectChildren(item, items)));
+    }
+
+    private getDirectProjectChildren(project: Project, projects: Project[]): Project[] {
+        return projects.filter((item: Project) => item.getParent() === project.getName());
+    }
+
+    private sortProjects(items: Project[]): Project[] {
+        const result: Project[] = [];
+
+        const defaultProject: Project = items.find((project: Project) => ProjectHelper.isDefault(project));
+
+        if (defaultProject) {
+            result.push(defaultProject);
+            result.push(...this.unwrapProjectChildren(defaultProject));
+        }
+
+        const level0projects: Project[] = items.filter(
+            (project: Project) => !ProjectHelper.isDefault(project) && this.projectLevel.get(project.getName()) === 0);
+        level0projects.forEach((level0project: Project) => {
+            result.push(level0project);
+            result.push(...this.unwrapProjectChildren(level0project));
+        });
+
```

Comment:
Hmm... Why do we have a special check for `defaultProject` and exclude if from `level0projects`  if we do the same operation (`result.push`, `this.unwrapProjectChildren`) for both?

Reply:
To ensure that Default project will always be first on the projects list

Reply:
Ok, then we can just sort `result` before returning it, to make sure the default project is first.

Reply:
We need to have default's chidren projects (with their children etc.) to be in the list right after default entry, so it is better this way - first process default with children, then the rest